src/modules/musig/keyagg_impl.h

```diff
@@ -0,0 +1,293 @@
+    secp256k1_gej pkj;
+    secp256k1_ge pkp;
+    size_t i;
+
+    VERIFY_CHECK(ctx != NULL);
+    if (agg_pk != NULL) {
+        memset(agg_pk, 0, sizeof(*agg_pk));
+    }
+    ARG_CHECK(pubkeys != NULL);
+    ARG_CHECK(n_pubkeys > 0);
+
+    ecmult_data.ctx = ctx;
+    ecmult_data.pks = pubkeys;
+
+    secp256k1_ge_set_infinity(&ecmult_data.second_pk);
+    for (i = 1; i < n_pubkeys; i++) {
+        if (secp256k1_memcmp_var(pubkeys[0], pubkeys[i], sizeof(*pubkeys[0])) != 0) {
+            secp256k1_ge pk;
+            if (!secp256k1_pubkey_load(ctx, &pk, pubkeys[i])) {
+                return 0;
+            }
+            ecmult_data.second_pk = pk;
+            break;
+        }
+    }
+
+    if (!secp256k1_musig_compute_pk_hash(ctx, ecmult_data.pk_hash, pubkeys, n_pubkeys)) {
+        return 0;
+    }
+    /* TODO: actually use optimized ecmult_multi algorithms by providing a
+     * scratch space */
+    if (!secp256k1_ecmult_multi_var(&ctx->error_callback, NULL, &pkj, NULL, secp256k1_musig_pubkey_agg_callback, (void *) &ecmult_data, n_pubkeys)) {
+        /* In order to reach this line with the current implementation of
+         * ecmult_multi_var one would need to provide a callback that can
+         * fail. */
+        return 0;
+    }
+    secp256k1_ge_set_gej(&pkp, &pkj);
+    secp256k1_fe_normalize_var(&pkp.y);
+    /* The resulting public key is infinity with negligible probability */
+    VERIFY_CHECK(!secp256k1_ge_is_infinity(&pkp));
+    if (keyagg_cache != NULL) {
+        secp256k1_keyagg_cache_internal cache_i = { 0 };
+        cache_i.pk = pkp;
+        cache_i.second_pk = ecmult_data.second_pk;
+        memcpy(cache_i.pk_hash, ecmult_data.pk_hash, sizeof(cache_i.pk_hash));
+        secp256k1_keyagg_cache_save(keyagg_cache, &cache_i);
+    }
+
+    secp256k1_extrakeys_ge_even_y(&pkp);
+    if (agg_pk != NULL) {
+        secp256k1_xonly_pubkey_save(agg_pk, &pkp);
+    }
+    return 1;
+}
+
+int secp256k1_musig_pubkey_get(const secp256k1_context* ctx, secp256k1_pubkey *agg_pk, const secp256k1_musig_keyagg_cache *keyagg_cache) {
+    secp256k1_keyagg_cache_internal cache_i;
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(agg_pk != NULL);
+    memset(agg_pk, 0, sizeof(*agg_pk));
+    ARG_CHECK(keyagg_cache != NULL);
+
+    if(!secp256k1_keyagg_cache_load(ctx, &cache_i, keyagg_cache)) {
+        return 0;
+    }
+    secp256k1_pubkey_save(agg_pk, &cache_i.pk);
+    return 1;
+}
+
+static int secp256k1_musig_pubkey_tweak_add_internal(const secp256k1_context* ctx, secp256k1_pubkey *output_pubkey, secp256k1_musig_keyagg_cache *keyagg_cache, const unsigned char *tweak32, int xonly) {
+    secp256k1_keyagg_cache_internal cache_i;
+    int overflow = 0;
+    secp256k1_scalar tweak;
+
+    VERIFY_CHECK(ctx != NULL);
+    if (output_pubkey != NULL) {
+        memset(output_pubkey, 0, sizeof(*output_pubkey));
+    }
+    ARG_CHECK(keyagg_cache != NULL);
+    ARG_CHECK(tweak32 != NULL);
```

Comment:
in https://github.com/bitcoin-core/secp256k1/commit/f8f4f3c7a10345295ed72cdae0faecff37a919e5:

Generally speaking, seems like the `ARG_CHECKS` should be happening in the public facing functions, not internal functions? i.e., doing a runtime check inside a function an external caller doesn't have access to seems a bit strange to me.

Reply:
At the cost of code duplication?

Reply:
In this case, I think it helps with the clarity to have the arg checking happen in the public function _before_ calling the internal function (even at the cost of code duplication). 

That being said, this is really just a style nit so feel free to resolve this discussion if you prefer the (current) more concise version.

Reply:
I tried this suggestion and I prefer the more concise version.